### PR TITLE
Add FXIOS-21302 TabPanelMiddlewareActionType tests

### PR DESCRIPTION
**Pull request by RRUIZ5**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-21302)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/21302)

## :bulb: Description
Adding test to `TabPanelStateTests`
Looking for feedback, so far I have only included tests for `TabPanelMiddlewareActionType`. After receiving feedback I will add the rest for `TabPanelViewActionType` 

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by RRUIZ5** [comment]
Just as I started working on this, there was a change made `TabsPanelState` which also included a bunch of tests for `TabScrollBehavior`. That is why the PR in its current state has conflicts.
My question to maintainers, are the tests that I was trying to make still worth it?

**Comment by cyndichin** [comment]
@RRUIZ5 thank you for picking up this ticket, apologies for the delay since I have been away. It seems the PR that added additional tests is [here](
https://github.com/mozilla-mobile/firefox-ios/pull/22596/files#diff-97176f5ebe537386b28c0a38f80fae74f240d38c5d58e84657593659a45d4b10
), which is specific to scrolling.

This ticket covers other actions not specific to scrolling, but I can look more closely when I'm back next week + review your specific tests in the PR. 

cc: @ih-codes

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabsPanelStateTests.swift
Comment (cyndichin) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabsPanelStateTests.swift:25]: separating these methods out makes it a bit harder to read these tests. What do you think about keeping some code so that it follows our given, when, then format and we can create helper methods for any setup needed that might bloat the test method?

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabsPanelStateTests.swift
Comment (cyndichin) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabsPanelStateTests.swift:121]: It seems that this is handled in `TabTrayState` and not in `TabsPanelState` now, so we don't need to add these tests here. As part of this ticket, we would like to increase coverage for actions that update the tabs panel state. 

There are still some areas that we can improve on despite the recent additions of tests to this class, see lack of coverage in the following screenshot:
image

Therefore, the tests you wrote for `refreshInactiveTabs` and `refreshTabs` and `willAppearTabPanel` are still relevant and helpful! Let me know if you have further question and can give another review when this is updated.

**Review by ih-codes — COMMENTED** [review]
Thanks for adding some new tests @RRUIZ5! I apologize for the delay in review as I was traveling.

@cyndichin basically covered everything already. Like she said, most of these tests are still good, but the 2 scroll related ones are now redundant (not to mention that the scroll logic has changed a decent bit overall).

---

For the two tests related to `didLoadTabPanel` and `didChangeTabPanel`, you can also assert for changes to the following state properties:

- `isPrivateMode`
- `inactiveTabs`
- `isInactiveTabsExpanded`

as those two actions cause state changes to more than just the `tabs` array ([see code here](https://github.com/mozilla-mobile/firefox-ios/blob/9f355b61cc7084b035c3627b8d6551a82b88beea/firefox-ios/Client/Frontend/Browser/Tabs/State/TabsPanelState.swift#L111-L115)). Thus, you can round out these tests a bit by checking additional properties on the `newState` you create with the `tabDisplayModel` passed into the action.

Also, you could enhance your check on the `tabs` array by asserting that it has an _equal size_ to your input tabs (`tabs.count`) rather than simply checking that it is not empty in the final line of your test. That will tell us more precisely that the `tabDisplayModel` properly updated the `TabPanelState`.

Thanks! Don't hesitate to reach out with additional questions, @cyndichin and I will help as best we can. Thanks for your work! 🙏

**Comment by RRUIZ5** [comment]
Thanks for all the feedback, I will update the PR

**Comment by RRUIZ5** [comment]
Thanks for your feedback @cyndichin  and @ih-codes, I have updated the PR, please let me know if there is something else I should update.

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabsPanelStateTests.swift
Comment (ih-codes) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabsPanelStateTests.swift:362]: Nice, this looks cleaner! 👌 

Is there a reason we cannot continue to use the preexisting `TabModel.emptyState()` function helper? It seems to accomplish the same thing here. I would like to continue using that instead of a new `MockTabModel()` function. Or at the very least, let's rename `MockTabModel` to follow swift function naming best practices. e.g. `generateTabModel` with lowercase first letter. I would hesitate to use the work `mock` in this function name as well since we do have a number of mock types in our project (e.g. MockProfile, MockRouter, etc.), but this function is in fact returning a standard (non-mocked) TabModel.

**Review by ih-codes — CHANGES_REQUESTED** [review]
Awesome work @RRUIZ5 thank you! I just have one small nitpick with your helper method. Other than that this is looking just about done. 👌

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabsPanelStateTests.swift
Comment (RRUIZ5) [reply firefox-ios/firefox-ios-tests/Tests/ClientTests/TabTray/TabsPanelStateTests.swift]: Thanks for the suggestion, I somehow missed that method.

**Review by cyndichin — APPROVED** [review]
thanks @RRUIZ5, looks great to me now!! I'll let @ih-codes confirm any issues on her side before running our CI / CD. appreciate the work and improving our tests 🥳

**Comment by ih-codes** [comment]
Our CI pipeline is not showing the build, so I need to close and reopen this this PR (sorry for the noise).

**Pull request merged**
